src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -415,32 +360,37 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
                     }
 
                     int chunkBytes = Math.Min(count, _sslState.MaxDataSize);
-                    int encryptedBytes;
-                    SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
-                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                    byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(chunkBytes + FrameOverhead);
+                    byte[] outBuffer = rentedBuffer;
+                    int encryptedBytes = 0;
+                    try
                     {
-                        // Re-handshake status is not supported.
-                        ProtocolToken message = new ProtocolToken(null, status);
-                        throw new IOException(SR.net_io_encrypt, message.GetException());
+                        SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
+                        if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                        {
+                            // Re-handshake status is not supported.
+                            ProtocolToken message = new ProtocolToken(null, status);
+                            throw new IOException(SR.net_io_encrypt, message.GetException());
+                        }
                     }
-
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    catch
```

Comment:
Other than the explicit throw a few lines up, what here might throw?  I'm wondering if we could just put the return just before the explicit throw above and get rid of the try/catch entirely.

Reply:
Yeah tracing through the code the only errors will be native and they are dealt with via error codes not exceptions